### PR TITLE
Simplify testing setup + Python 3.12 tests

### PR DESCRIPTION
**Pull request by dstansby**
I got a bit distracted by wanting to add Python 3.12 testing here 😄  This generally cleans up the testing infastructure by:

- Using a test matrix in GitHub Actions
- Using https://github.com/ymyzk/tox-gh-actions to reduce boilerplate config in `tox.ini` and the GitHub action configuration

I also noticed `fsspec==2023.6.0` was pinned for tests, but newer versions of `fsspec` break the tests. So I've pinned `fsspec` in `setup.py` correctly.

### CODE REVIEW

**Review by will-moore — APPROVED** [review]
Looks good, Thanks.
We now have `18 Checks` for this PR at https://github.com/ome/ome-zarr-py/pull/338/checks compared with 12 at previously merged PR
https://github.com/ome/ome-zarr-py/pull/326/checks so more complete coverage.

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -4,67 +4,30 @@ on: [push, pull_request]
...
-          - {os: windows-latest, python_Version: '3.8', toxenv: 'py38'}
-          - {os: windows-latest, python_Version: '3.9', toxenv: 'py39'}
-          - {os: ubuntu-latest, python_Version: '3.11', toxenv: 'py311'}
-          - {os: ubuntu-latest, python_Version: '3.10', toxenv: 'py310'}
-          - {os: ubuntu-latest, python_Version: '3.8', toxenv: 'py38'}
-          - {os: ubuntu-latest, python_Version: '3.9', toxenv: 'py39'}
-          - {os: macos-latest, python_Version: '3.11', toxenv: 'py311'}
-          - {os: macos-latest, python_Version: '3.10', toxenv: 'py310'}
-          - {os: macos-latest, python_Version: '3.8', toxenv: 'py38'}
-          # missing numcodecs wheels on 3.9. conda not yet an option. see gh-51
```
Comment (joshmoore) [inline .github/workflows/build.yml:24]: :+1: if this is now all fixed. Thanks.

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -4,67 +4,30 @@ on: [push, pull_request]
...
+      - uses: actions/checkout@v4
 
-      - uses: actions/setup-python@v4
-        name: Install Python ${{ matrix.python_version }}
+      - name: Set up Python ${{ matrix.python-version }}
+        uses: actions/setup-python@v4
         with:
-          python-version: ${{ matrix.python_version }}
-
-      - name: Install ubuntu libraries
```
Comment (joshmoore) [inline .github/workflows/build.yml:35]: I'm concerned that by skipping these we're now omitting a test. Or did those tests get moved to napari-ome-zarr, @will-moore?

File: tox.ini [hunk 3]
```diff
@@ -1,26 +1,18 @@
...
 
-[testenv:py310-coverage]
-passenv =
-    CI
-    GITHUB_ACTIONS
-usedevelop = true
 commands =
-    pytest --cov-report=xml --cov=./ome_zarr  --cov-append {posargs:-v}
-    codecov -f codecov.xml
+    pytest --cov-report=xml --cov=./ome_zarr  --cov-append {posargs}
```
[the remark this answers is not in the document]
Comment (dstansby) [reply tox.ini:18]: No, codecov is happy recieving multiple coverage reports from different test runs.

File: tox.ini [hunk 4]
```diff
@@ -1,26 +1,18 @@
...
-# in multiple virtualenvs. This configuration file will run the
-# test suite on all supported python versions. To use it, "pip install tox"
-# and then run "tox" from this directory.
-
 [tox]
-envlist = py{38,39,310,311}
+envlist = py{38,39,310,311,312}
+
+
+[gh-actions]
```
[the remark this answers is not in the document]
Comment (dstansby) [reply tox.ini:5]: It's completely usable locally - the only thing the `gh-actions` bit does is tell `tox` which environment to run on each GitHub actions runner, depeneding on the version of Python installed.

File: pytest.ini [hunk 5]
```diff
@@ -1,2 +0,0 @@
-[pytest]
-addopts = --doctest-modules
```
[the remark this answers is not in the document]
Comment (dstansby) [reply pytest.ini:2]: Woops, shouldn't have removed this.

File: CHANGELOG.md [hunk 6]
```diff
@@ -1,3 +1,8 @@
+# 0.8.4 (unreleased)
+
+- Correctly maximum compatible fsspec version. ([#338](https://github.com/ome/ome-zarr-py/pull/338))
```
Comment (dstansby) [inline CHANGELOG.md:3]:
```suggestion
- Correctly specify maximum compatible fsspec version. ([#338](https://github.com/ome/ome-zarr-py/pull/338))
```

File: setup.py [hunk 7]
```diff
@@ -19,7 +19,7 @@ def read(fname):
 install_requires += (["dask"],)
 install_requires += (["distributed"],)
 install_requires += (["zarr>=2.8.1"],)
-install_requires += (["fsspec[s3]>=0.8,!=2021.07.0,!=2023.09.0"],)
+install_requires += (["fsspec[s3]>=0.8,!=2021.07.0,<2023.09.0"],)
```
[the remark this answers is not in the document]
Comment (joshmoore) [reply setup.py:22]: Thank! Taken care of now.

**Pull request merged**